framework/src/wifi_manager/slsi_wifi_utils_lwnl80211.c

```diff
@@ -0,0 +1,483 @@
+				}
+				free_scan_data(scan_list);
+			} else {
+				g_cbk.scan_done(WIFI_UTILS_SUCCESS, scan_list, NULL);
+				free(&(scan_list->ap_info));
+			}
+			break;
+		}
+		default:
+			WU_LOG("Bad status received (%d)\n", msg.status);
+			WU_ERR;
+			break;
+		}
+	}
+
+	pthread_exit(NULL);
+	return NULL;
+}
+
+wifi_utils_result_e wifi_utils_init(void)
+{
+	WU_ENTER;
+
+	pthread_attr_t p_attr;
+	int status;
+	struct mq_attr attr;
+	wifi_utils_result_e ret = WIFI_UTILS_FAIL;
+
+	int fd = open(LWNL80211_PATH, O_RDWR);
+	WU_CHECK_ERR(fd);
+
+	attr.mq_maxmsg = LWNL80211_MQUEUE_MAX_DATA_NUM;
+	attr.mq_msgsize = sizeof(lwnl80211_cb_data);
+	attr.mq_flags = 0;
+	attr.mq_curmsgs = 0;
+
+	snprintf(g_hnd.mqname, sizeof(g_hnd.mqname), "%01x", (unsigned long)((uintptr_t)&g_hnd));
+	g_hnd.mq = mq_open(g_hnd.mqname, O_RDWR | O_CREAT, 0666, &attr);
+	if (g_hnd.mq == NULL) {
+		close(fd);
+		WU_LOG("Failed to open mq\n");
+		return WIFI_UTILS_FAIL;
+	}
+
+	/* Start to send ioctl */
+	lwnl80211_data data_in = {NULL, 0, 0};
+	data_in.data_len = sizeof(g_hnd.mqname);
+	data_in.data = (mqd_t *)malloc(data_in.data_len);
+	if (!data_in.data) {
+		WU_LOG("Failed to alloc lw80211_data input\n");
+		ret = WIFI_UTILS_FAIL;
+		goto errout;
+	}
+	memcpy(data_in.data, g_hnd.mqname, data_in.data_len);
+
+	status = pthread_attr_init(&p_attr);
+	if (status != 0) {
+		WU_ERR;
+		return WIFI_UTILS_FAIL;
+	}
+
+	status = pthread_create(&g_hnd.cb_receiver, &p_attr, wifi_utils_callback_handler, NULL);
+	if (status != 0) {
+		WU_ERR;
+		return WIFI_UTILS_FAIL;
+	}
+	pthread_setname_np(g_hnd.cb_receiver, "lwnl80211 cb handler");
+
+	WU_CALL_ERROUT(fd, g_hnd, LWNL80211_REGISTERMQ, data_in);
+	free(data_in.data);
+	data_in.data = NULL;
+	data_in.data_len = 0;
+	data_in.res = 0;
+
+	WU_CALL_ERROUT(fd, g_hnd, LWNL80211_INIT, data_in);
+
+	WU_CLOSE(fd);
+	return WIFI_UTILS_SUCCESS;
+
+errout:
+	WU_CLOSE(fd);
+	mq_close(g_hnd.mq);
+	g_hnd.mq = NULL;
+	if (data_in.data) {
+		free(data_in.data);
+	}
+	return ret;
+}
+
+wifi_utils_result_e wifi_utils_deinit(void)
+{
+	WU_ENTER;
+
+	int status;
+
+	int fd = open(LWNL80211_PATH, O_RDWR);
+	WU_CHECK_ERR(fd);
+
+	lwnl80211_data data_in = {NULL, 0, 0};
+
+	WU_CALL(fd, g_hnd, LWNL80211_UNREGISTERMQ, data_in);
+
+#ifndef CONFIG_DISABLE_SIGNALS
+	pthread_kill(g_hnd.cb_receiver, 9);
+	usleep(HALF_SECOND_USEC_USEC);
+#endif
+
+	status = pthread_cancel(g_hnd.cb_receiver);
+	if (status == ESRCH) {
+		WU_LOG("Callback receiver thread already closed successfully\n");
+	} else if (status != 0) {
+		WU_ERR;
+		WU_CLOSE(fd);
+		return WIFI_UTILS_FAIL;
+	}
+	pthread_join(g_hnd.cb_receiver, NULL);
+
+	g_cbk = (wifi_utils_cb_s){NULL, NULL, NULL, NULL, NULL};
+
+	WU_CALL(fd, g_hnd, LWNL80211_DEINIT, data_in);
+
+	WU_CLOSE(fd);
+	mq_close(g_hnd.mq);
+	g_hnd.mq = NULL;
+
+	return WIFI_UTILS_SUCCESS;
+}
+
+wifi_utils_result_e wifi_utils_scan_ap(void *arg)
+{
+	WU_ENTER;
+
+	int fd = open(LWNL80211_PATH, O_RDWR);
```

Comment:
is it fine to declare and int initialize variable ' fd ' like this?

Reply:
It is fine, as far as I know. Thanks.